sdk/storage/Azure.Storage.Common/src/Shared/WriteStream.cs

```diff
@@ -0,0 +1,181 @@
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => throw new NotSupportedException();
+
+        public override long Seek(long offset, SeekOrigin origin)
+            => throw new NotSupportedException();
+
+        public override void SetLength(long value)
+            => throw new NotSupportedException();
+
+        public override void Write(
+            byte[] buffer,
+            int offset,
+            int count)
+            => WriteInternal(
+                async: false,
+                buffer,
+                offset,
+                count,
+                cancellationToken: default)
+            .EnsureCompleted();
+
+        public override async Task WriteAsync(
+            byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+            => await WriteInternal(
+                async: true,
+                buffer,
+                offset,
+                count,
+                cancellationToken)
+            .ConfigureAwait(false);
+
+        protected virtual async Task WriteInternal(
+            bool async,
+            byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            ValidateWriteParameters(buffer, offset, count);
+            int remaining = count;
+
+            // New bytes will fit in the buffer.
+            if (count <= _buffer.Capacity - _buffer.Position)
+            {
+                await WriteToBuffer(async, buffer, offset, count, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // Flush the buffer.
+                await FlushInternal(async, cancellationToken).ConfigureAwait(false);
+
+                while (remaining > 0)
+                {
+                    await WriteToBuffer(
+                        async,
+                        buffer,
+                        offset,
+                        Math.Min(remaining, _buffer.Capacity),
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Renaming bytes won't fit in buffer.
+                    if (remaining > _buffer.Capacity)
+                    {
+                        await FlushInternal(async, cancellationToken).ConfigureAwait(false);
+                        remaining -= _buffer.Capacity;
+                        offset += _buffer.Capacity;
+                    }
+
+                    // Remaining bytes will fit in buffer.
+                    else
+                    {
+                        remaining = 0;
+                    }
+                }
+            }
+            _position += count;
+        }
+
+        protected virtual async Task FlushInternal(bool async, CancellationToken cancellationToken)
+            => await AppendInternal(async, cancellationToken).ConfigureAwait(false);
```

Comment:
This seems to make strong assumption about what derived types do (optimizing for append blob).
I'd rather force derived type to define this.

Also, In general, consider taking delegates for append and flush instead of creating subtypes (we do that in uploader). Such implementation is more open for future changes and allows more variations between blob types when they come in the future.

Reply:
3/5 clients (Page Blob, Append Blob, and Share File) have this Flush behavior.  Only  Block Blob and Data Lake File have a separate Flush method.  I was attempting to consolidate as much logic as possible in the WriteStream.

I think I like the current design better than using delegates, as I did in Open Write because this seems simpler and easier to understand.

Reply:
What if we introduce few more blobs that don't have this behavior and balance will shift? I think implementation detail leak is worse than having to alias flush to append in 3 classes. In general `WriteStream` being abstract shouldn't be aware of types of blobs. It should say that "I don't know how to Append" and "I don't know how to Flush" and force derivatives to fill in these behaviors.
Other way to see that is : I need to jump between Append/Page Blob and WriteStream to understand full picture.
Yet another way to see that is: If new blob is introduced someone can inherit this behavior without giving it much thought. Making Flush abstract would force to think about that.

For delegates vs inheritance there's quite a lot of literature why inheritance is not best choice for code sharing:
https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html
https://en.wikipedia.org/wiki/Composition_over_inheritance
Composition based solution usually allow easier evolution (how easy it is to add functionality over time).

Reply:
I'm fine making Flush abstract.

Can you point me to the delegates in the uploader?  I'm not seeing them in [PartitionedUploader](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/PartitionedUploader.cs), I think I'm looking in the wrong place.

I think it's super unlikely that we will introduce that doesn't follow the Append or Append + Flush patterns.  It would have to be a 3 step process to add new data or something.

I can look into passing around a bunch of lamdbas.

Reply:
Made flush abstract in the next iteration.